src/devices/cpu/e132xs/e132xsop.hxx

```diff
@@ -1311,191 +1250,248 @@ void hyperstone_device::hyperstone_add_global_global()
+template <hyperstone_device::reg_bank DST_GLOBAL, hyperstone_device::reg_bank SRC_GLOBAL>
+void hyperstone_device::hyperstone_andn()
+{
+	check_delay_PC();
 
-	dreg += sreg;
-	m_local_regs[dst_code] = dreg;
+	const uint32_t dst_code = DST_GLOBAL ? DST_CODE : ((DST_CODE + GET_FP) & 0x3f);
+	const uint32_t sreg = SRC_GLOBAL ? m_global_regs[SRC_CODE] : m_local_regs[(SRC_CODE + GET_FP) & 0x3f];
+	const uint32_t dreg = (DST_GLOBAL ? m_global_regs : m_local_regs)[dst_code] & ~sreg;
+	if (DST_GLOBAL)
+		set_global_register(dst_code, dreg);
+	else
+		m_local_regs[dst_code] = dreg;
 
 	if (dreg == 0)
 		SR |= Z_MASK;
-	SR |= SIGN_TO_N(dreg);
+	else
+		SR &= ~Z_MASK;
 
 	m_icount -= m_clock_cycles_1;
 }
 
-void hyperstone_device::hyperstone_add_local_local()
+template <hyperstone_device::reg_bank DST_GLOBAL, hyperstone_device::reg_bank SRC_GLOBAL>
+void hyperstone_device::hyperstone_or()
 {
 	check_delay_PC();
 
-	const uint32_t fp = GET_FP;
-	const uint32_t dst_code = (DST_CODE + fp) & 0x3f;
-
-	const uint32_t sreg = m_local_regs[(SRC_CODE + fp) & 0x3f];
-	uint32_t dreg = m_local_regs[dst_code];
+	const uint32_t dst_code = DST_GLOBAL ? DST_CODE : ((DST_CODE + GET_FP) & 0x3f);
+	const uint32_t sreg = SRC_GLOBAL ? m_global_regs[SRC_CODE] : m_local_regs[(SRC_CODE + GET_FP) & 0x3f];
+	const uint32_t dreg = (DST_GLOBAL ? m_global_regs : m_local_regs)[dst_code] | sreg;
+	if (DST_GLOBAL)
+		set_global_register(dst_code, dreg);
+	else
+		m_local_regs[dst_code] = dreg;
 
-	const uint64_t tmp = (uint64_t)sreg + (uint64_t)dreg;
+	if (dreg == 0)
+		SR |= Z_MASK;
+	else
+		SR &= ~Z_MASK;
 
-	SR &= ~(C_MASK | V_MASK | Z_MASK | N_MASK);
+	m_icount -= m_clock_cycles_1;
+}
 
-	SR |= (tmp & 0x100000000L) >> 32;
-	SR |= ((sreg ^ tmp) & (dreg ^ tmp) & 0x80000000) >> 28;
+template <hyperstone_device::reg_bank DST_GLOBAL, hyperstone_device::reg_bank SRC_GLOBAL>
+void hyperstone_device::hyperstone_xor()
+{
+	check_delay_PC();
 
-	dreg += sreg;
-	m_local_regs[dst_code] = dreg;
+	const uint32_t dst_code = DST_GLOBAL ? DST_CODE : ((DST_CODE + GET_FP) & 0x3f);
+	const uint32_t sreg = SRC_GLOBAL ? m_global_regs[SRC_CODE] : m_local_regs[(SRC_CODE + GET_FP) & 0x3f];
+	const uint32_t dreg = (DST_GLOBAL ? m_global_regs : m_local_regs)[dst_code] ^ sreg;
+	if (DST_GLOBAL)
+		set_global_register(dst_code, dreg);
+	else
+		m_local_regs[dst_code] = dreg;
 
 	if (dreg == 0)
 		SR |= Z_MASK;
-	SR |= SIGN_TO_N(dreg);
+	else
+		SR &= ~Z_MASK;
 
 	m_icount -= m_clock_cycles_1;
 }
 
-void hyperstone_device::hyperstone_adds_global_global()
+
+
+template <hyperstone_device::reg_bank DST_GLOBAL, hyperstone_device::reg_bank SRC_GLOBAL>
+void hyperstone_device::hyperstone_subc()
 {
-	const uint32_t src_code = SRC_CODE;
-	const uint32_t dst_code = DST_CODE;
-	const int32_t sreg = (src_code == SR_REGISTER) ? (SR & C_MASK) : (int32_t)m_global_regs[src_code];
-	const int32_t dreg = m_global_regs[dst_code];
-	const int64_t tmp = (int64_t)sreg + (int64_t)dreg;
+	const uint32_t fp = GET_FP;
```

Comment:
There's no call to `check_delay_PC` here - the similar `addc` handler does have such a call. Is this intentional or an oversight?

Reply:
Any ALU opcode that fails to check check_delay_PC is oversight on my part and should be added in the relevant place (after any decode calls which increment PC).